File: src/main/java/com/key/oa/config/DesensitizeJsonSerializer.java
package com.key.oa.config;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.BeanProperty;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.ser.ContextualSerializer;
import com.key.oa.annotation.Desensitize;

import java.io.IOException;
import java.util.Objects;

/**
 * 自定义的JsonSerializer
 *
 * @author 孙强
 */
public class DesensitizeJsonSerializer extends JsonSerializer<String> implements ContextualSerializer {
    private Desensitize.DesensitizeStrategy strategy;


    @Override
    public void serialize(String value, JsonGenerator gen, SerializerProvider serializers) throws IOException {
        gen.writeString(strategy.desensitizeSerializer().apply(value));
    }

    @Override
    public JsonSerializer<?> createContextual(SerializerProvider prov, BeanProperty property) throws JsonMappingException {
        Desensitize annotation = property.getAnnotation(Desensitize.class);
        if (Objects.nonNull(annotation)&&Objects.equals(String.class, property.getType().getRawClass())) {
            this.strategy = annotation.strategy();
            return this;
        }
        return prov.findValueSerializer(property.getType(), property);
    }
}

File: src/main/java/com/key/oa/config/DesensitizeJsonSerializer.kt
package com.key.oa.config

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.BeanProperty
import com.fasterxml.jackson.databind.JsonSerializer
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.ser.ContextualSerializer
import com.key.oa.annotation.Desensitize
import com.key.oa.annotation.Desensitize.DesensitizeStrategy

class DesensitizeJsonSerializer : JsonSerializer<String>(), ContextualSerializer {
    private lateinit var strategy: DesensitizeStrategy

    override fun serialize(value: String, gen: JsonGenerator, serializers: SerializerProvider) {
        gen.writeString(strategy.desensitizeSerializer().apply(value))
    }

    override fun createContextual(prov: SerializerProvider, property: BeanProperty): JsonSerializer<*> {
        val annotation: Desensitize? = property.getAnnotation(Desensitize::class.java)
        if (annotation !== null && property.type.rawClass == String::class.java) {
            this.strategy = annotation.strategy
            return this
        }
        return prov.findValueSerializer(property.type, property)
    }
}